### PR TITLE
Fix default post_date format + add date format magic tag options

### PR DESCRIPTION
**Pull request by JoryHogeveen**
Fixes: #5675
Fixes: #4525

## Description

THIS PR IS NOT DONE.

The relates issue noted that using `{@post_date}` always displays the mysql format (Y-m-d h:i).
This PR sets this to the default formatting settings from WordPress.

This PR also adds a format param to the magic tag: `{@post_date._format.Y-m-d}`, `{@post_date._format.Y m d}` and even `{@post_date._format.Y.m.d}` (note the dots).
Currently this PR only supports this for core-field dates (`post_date`). But might be a cool feature for other date fields as well.

**Would like to have feedback/ideas on how to best implement this.**

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style.
- [ ] My code follows has proper inline documentation.
- [ ] My code includes PHP Unit Tests (if applicable)

### CODE REVIEW

**Comment by patrikhuber** [comment]
Hi @JoryHogeveen,

Cool, this would be great, thank you for moving this forward! Big thumbs up from me on the feature, and by just googling "wordpress pods date format", there are dozens of users looking for something like it.

As a programmer myself, I would perhaps suggest to put the format string in parentheses? I would say that it gets a bit dodgy with whitespaces otherwise, like `{@post_date._format.Y m d}`. For example `{@post_date._format("Y m d")}`. But since everything is inside curly braces `{...}`, even your whitespace example isn't too unclear.

**Comment by JoryHogeveen** [comment]
@patrikhuber 
Thanks. And about the parentheses, that isn't compatible with the Pods magic tag formats, so even if it would be enclosed with quotes, there won't be any function-like formatting here.

File: classes/Pods.php [hunk 1]
```diff
@@ -1143,9 +1137,52 @@ public function field( $name, $single = null, $raw = false ) {
...
+									if ( $field ) {
+										$format = implode( '.', $field );
+										$value  = date_i18n( $format, strtotime( $value ) );
+
+										$field_data['options']['datetime_type']          = 'custom';
+										$field_data['options']['datetime_time_type']     = 'custom';
+										$field_data['options']['datetime_format_custom'] = $format;
+									} else {
+										$field_data['options']['datetime_type']          = 'wp';
+										$field_data['options']['datetime_time_type']     = 'wp';
```
Comment (sc0ttkclark) [inline classes/Pods.php:1161]:
```suggestion
										$field_data['options']['datetime_type']      = 'wp';
										$field_data['options']['datetime_time_type'] = 'wp';
```

File: classes/Pods.php [hunk 2]
```diff
@@ -1143,9 +1137,52 @@ public function field( $name, $single = null, $raw = false ) {
...
+				foreach ( $date_fields as $date_field ) {
+					if (
+						0 === strpos( $params->name, $date_field . '.' )
+					) {
+						$field = explode( '.', $params->name );
+
+						if ( 1 < count( $field ) ) {
+							array_shift( $field ); // Remove field name.
+							$type = array_shift( $field );
+							switch ( $type ) {
```
Comment (sc0ttkclark) [inline classes/Pods.php:1150]:
```suggestion

							switch ( $type ) {
```

File: classes/Pods.php [hunk 3]
```diff
@@ -1143,9 +1137,52 @@ public function field( $name, $single = null, $raw = false ) {
...
 				} //end foreach
+
+				foreach ( $date_fields as $date_field ) {
+					if (
+						0 === strpos( $params->name, $date_field . '.' )
+					) {
+						$field = explode( '.', $params->name );
+
+						if ( 1 < count( $field ) ) {
+							array_shift( $field ); // Remove field name.
```
Comment (sc0ttkclark) [inline classes/Pods.php:1148]:
```suggestion
							// Remove field name.
							array_shift( $field );

```

**Comment by sc0ttkclark** [comment]
Looks good, just needs tests

**Comment by JoryHogeveen** [comment]
@sc0ttkclark 
For tests we might want to merge https://github.com/pods-framework/pods/pull/4530 first. Maybe merge it in this PR?

File: classes/Pods.php [hunk 4]
```diff
@@ -1676,6 +1722,7 @@ public function field( $name, $single = null, $raw = false ) {
 									foreach ( $data as $item_id => $item ) {
 										// $field is 123x123, needs to be _src.123x123
 										$full_field = implode( '.', array_splice( $params->traverse, $key ) );
+										var_dump( $full_field );
```
Comment (sc0ttkclark) [inline classes/Pods.php:1725]:
```suggestion
```

**Comment by JoryHogeveen** [comment]
@sc0ttkclark
I found a little glitch in our usage of dates.
We set the date fields to be `datetime`, which is logical. But, as the tests from @jamesgol compared the `post_date` tag with `get_the_date()` this will not match since `get_the_date()` only returns the date, not the time.

I can think of various options to add other magic tags:
- `{@post_date}` > Only date in default WP format
- `{@post_time}` > Only time in default WP format
- `{@post_datetime}` > Date & time in default WP format
etc.

Adding these tags would mean I'd need to extend the WP object fields for posts in PodsAPI.

But we can also leave it like this since it now supports custom formats: `{@post_date._format.##`

What do you think would be best?

**Comment by sc0ttkclark** [comment]
We probably need to choose something different because post_date (data-wise) is really the post date+time. Maybe something like the_date or the_datetime or the_time?

**Comment by JoryHogeveen** [comment]
@sc0ttkclark 
Sounds good to me as well. So with this `pods_date` remains unchanged?
I can refactor this PR to do that. But we'll need to create docs for this so people know `post_date` is the actual database value and `the_##` is the display value.

**Comment by sc0ttkclark** [comment]
@JoryHogeveen Yes, I think you meant `post_date` not `pods_date`, but yes that value remains unchanged. `the_*` variations would do display.

**Comment by sc0ttkclark** [comment]
All of the logic will need to be reworked here in the context of the new field value mapping added in Pods 2.8. Also overriding options in the field config means that it modifies the object and that could be a problem for tests etc.

Overall, I think we need to look at a better solution for modifiers outside of the regular field mapping in 2.8, maybe something that can run after the value itself is retrieved and can change the resulting value/display.

**Comment by dmkjr** [comment]
Any movement on this?  Looking for a method to only include the Month/Year in our PODS template.

**Pull request closed without merging**